src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -137,25 +157,36 @@ where
 pub async fn convert_all_bits<F, C, S>(
     ctx: &C,
     locally_converted_bits: &[Vec<BitConversionTriple<S>>],
-) -> Result<Vec<Vec<S>>, Error>
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<Vec<Vec<Vec<S>>>, Error>
 where
     F: Field,
     C: Context<F, Share = S>,
     S: ArithmeticSecretSharing<F>,
 {
-    let futures = locally_converted_bits
-        .iter()
-        .enumerate()
-        .map(|(bit_num, one_column)| {
-            convert_bit_list(
-                ctx.narrow(&ModulusConversion(bit_num.try_into().unwrap()))
-                    .set_total_records(one_column.len()),
-                one_column,
-            )
-        })
-        .collect::<Vec<_>>();
-    let converted_shares = try_join_all(futures).await?;
-    Ok(converted_shares)
+    let num_records = locally_converted_bits.len();
+    try_join_all(
+        (0..num_bits)
+            .step_by(num_multi_bits.try_into().unwrap())
+            .map(|bit_num| {
+                let last_bit = std::cmp::min(num_multi_bits + bit_num, num_bits) as usize;
+                try_join_all(
+                    locally_converted_bits
+                        .iter()
+                        .enumerate()
+                        .map(|(idx, record)| {
+                            convert_bit_list(
+                                ctx.narrow(&ModulusConversion(bit_num))
+                                    .set_total_records(num_records),
```

Comment:
I feel like we should pull these lines out of the inner loop since they do not depend on the record at all. Right now you're calling narrow num_records times, when we only need to call it once and from then on just clone the context.

Reply:
Agreed, changing this